src/mbedtls/engine.c

```diff
@@ -287,6 +297,42 @@ static int internal_cert_verify(void *ctx, mbedtls_x509_crt *crt, int depth, uin
         }
     }
 
+#if defined(__APPLE__)
+    if (*flags & MBEDTLS_X509_BADCERT_NOT_TRUSTED) {
+        CFMutableArrayRef certs = CFArrayCreateMutable(kCFAllocatorDefault, 1, NULL);
+        mbedtls_x509_crt *c1 = crt;
+        while(c1) {
+            CFDataRef raw = CFDataCreate(kCFAllocatorDefault, c1->raw.p, (CFIndex)c1->raw.len);
+            SecCertificateRef c = SecCertificateCreateWithData(kCFAllocatorDefault, raw);
+
+            CFArrayAppendValue(certs, c);
+            c1 = c1->next;
+        }
+
+        SecPolicyRef x509policy = SecPolicyCreateBasicX509();
+        SecTrustRef trust;
+        OSStatus status = SecTrustCreateWithCertificates(certs, x509policy, &trust);
+        if (status == errSecSuccess) {
+            CFErrorRef err = 0;
+            if (SecTrustEvaluateWithError(trust, &err)) {
+                *flags &= ~MBEDTLS_X509_BADCERT_NOT_TRUSTED;
+            } else {
+                CFStringRef e = CFErrorCopyDescription(err);
+                char errbuf[1024];
+                CFStringGetCString(e, errbuf, 1024, kCFStringEncodingUTF8);
+                UM_LOG(WARN, "certificate verify failed: %s", errbuf);
+            }
+            CFRelease(trust);
+        } else {
+            CFStringRef error = SecCopyErrorMessageString(status, NULL);
+            char err[128];
+            CFStringGetCString(error, err, 128, kCFStringEncodingASCII);
+            UM_LOG(WARN, "failed to create Trust object: %s", err);
+        }
+        CFRelease(certs);
```

Comment:
releases all the data for each cert appended?

Reply:
added a few more `CFRelease` calls and passed proper callbacks to `CFArray`